compiler/parser/RdrHsSyn.hs

```diff
@@ -2304,127 +2384,97 @@ classInstDecl tyApps funBinds =
+  -> LHsExpr GhcPs               -- function body
   -> Maybe (LHsLocalBinds GhcPs) -- local binds
-  -> P (Maybe (LHsBind GhcPs))   -- function binding
-mkMaintainerFunBindDecl _ _ Nothing _ = return Nothing
-mkMaintainerFunBindDecl fname _con_name (Just body) binds = do
-  let tag = noLoc $ mkRdrUnqual (mkVarOcc fname)
-      key = XPat (noLoc $ VarPat noExt $ noLoc $ mkRdrUnqual (mkVarOcc "key"))
+  -> LHsBind GhcPs               -- function binding
+mkTemplateClassMethod rawMethodName args body mBinds = do
+  let fullMethodName = noLoc $ mkRdrName rawMethodName
+      ctx = matchContext fullMethodName
       loc = getLoc body
-      ctx = matchContext tag
-      match = matchWithBinds ctx [key] loc body binds
+      binds = fromMaybe (noLoc emptyLocalBinds) mBinds
+      match = matchWithBinds ctx args loc body binds
       match_group = matchGroup loc match
-  funBind loc tag match_group
-
--- [Controlled choice groups - Old syntax] This data structure is a
--- tiny backwards compatibility hack for 'controller' function
--- bindings.
-data ArgPattern = ArgWildcardPat | ArgAsPat
-
--- | Construct a 'controller' function binding.
-mkTemplateControllerFunBindDecl
-  :: Located RdrName             -- data ctor 'T'
-  -> Maybe (LHsExpr GhcPs)       -- function body
-  -> Located RdrName             -- data ctor 'S'
-  -> Maybe (LHsType GhcPs)       -- record fields
-  -> ArgPattern                  -- 'arg@S{..}' or '_'?
-  -> Maybe (LHsLocalBinds GhcPs) -- local binds
-  -> P (Maybe (LHsBind GhcPs))   -- function binding
-mkTemplateControllerFunBindDecl _ Nothing _  _ _ _ = return Nothing
-mkTemplateControllerFunBindDecl
-  conName (Just body) choiceConName mbChoiceFields argPatType binds = do
-  let tag = noLoc $ mkRdrUnqual (mkVarOcc "choiceController")
+  funBind loc fullMethodName match_group
+
+-- | Prefix the names of methods in `class XInstance` to avoid clashes.
+prefixTemplateClassMethod :: String -> String
+prefixTemplateClassMethod = (++) "_"
+
+mkMagic :: String -> LHsExpr GhcPs
+mkMagic methodName =
+  let mkVarQual :: OccName -> LHsExpr GhcPs = noLoc . HsVar noExt . noLoc . qualifyDesugar
+      mkAppType :: LHsExpr GhcPs -> LHsType GhcPs -> LHsExpr GhcPs
+      mkAppType e ty = noLoc $ HsAppType noExt e (mkHsWildCardBndrs ty)
+  in  mkAppType (mkVarQual $ mkVarOcc "magic") $
+        noLoc $ HsTyLit noExt $ HsStrTy NoSourceText $ mkFastString methodName
+
+mkTemplateChoiceMethods ::
+     Located RdrName             -- template data ctor 'T'
+  -> LHsLocalBinds GhcPs         -- local binds
+  -> CombinedChoiceData          -- choice data and controllers
+  -> [LHsBind GhcPs]             -- function binding
+mkTemplateChoiceMethods conName binds (CombinedChoiceData controllers ChoiceData{..} flexible) =
+  let templateName = occNameString $ rdrNameOcc $ unLoc conName
+      choiceName = occNameString $ rdrNameOcc $ unLoc cdChoiceName
+      self = XPat $ noLoc $ VarPat noExt $ noLoc $ mkRdrName "self"
```

Comment:
I think this is worthy of a utility like `asPatRecWild` (just to group these things together)